releasing/plugin-commands-deploy/test/shared-lockfile.test.ts

```diff
@@ -960,3 +961,76 @@ test('deploy with a shared lockfile that has peer dependencies suffix in workspa
     expect(readPackageJson(`deploy/node_modules/.pnpm/${name}/node_modules/project-2`)).toStrictEqual(preparedManifests['project-2'])
   }
 })
+
+test('deploy with a shared lockfile should keep files created by lifecycle scripts', async () => {
+  const preparedManifests: Record<string, ProjectManifest> = {
+    root: {
+      name: 'root',
+      version: '0.0.0',
+      private: true,
+      pnpm: {
+        neverBuiltDependencies: [],
+      },
+    },
+    'project-0': {
+      name: 'project-0',
+      version: '0.0.0',
+      dependencies: {
+        '@pnpm.e2e/install-script-example': '*',
+      },
+    },
+  }
+
+  preparePackages([
+    {
+      location: '.',
+      package: preparedManifests.root,
+    },
+    preparedManifests['project-0'],
+  ])
+  await writeYamlFile('pnpm-workspace.yaml', { packages: ['project-0', '!store/**'] })
+
+  const {
+    allProjects,
+    allProjectsGraph,
+    selectedProjectsGraph,
+  } = await filterPackagesFromDir(process.cwd(), [{ namePattern: 'project-0' }])
+
+  await install.handler({
```

Comment:
@KSXGitHub could you help me with this test case also?

Here it fails already with `Cannot have both neverBuiltDependencies and onlyBuiltDependencies` - but it don't specify `onlyBuiltDependencies` anywhere?

Reply:
`getOptionsFromRootManifest` was called twice.

The first time with `manifestDir = '', manifest = {}, neverBuiltDependencies = undefined, onlyBuiltDependencies = []`. The result was passed as `opts: RecursiveOptions` to the function `recursive` in file `recursive.ts`.

The second time with `manifestDir = <full path>, manifest = <your root package>, neverBuiltDependencies = [], onlyBuiltDependencies = undefined`. This called was at line 134 (`...getOptionsFromRootManifest(rootManifestDir, manifestsByPath[rootManifestDir]?.manifest ?? {}),`).

The result of the first time and the second time was merged into the object `installOpts`, resulting in both properties being an empty array.

Reply:
https://github.com/pnpm/pnpm/pull/8990#issuecomment-2614513135